### PR TITLE
Handle promise-proxy deprecation

### PR DESCRIPTION
**Pull request by Monkeychip**
### Description
Two areas where the promise-proxy errors [See [ember RFC](https://rfcs.emberjs.com/id/0846-ember-data-deprecate-proxies/) for details on this deprecation] showed up:

1. Transform iterated through a loop of unresolved promises. I addressed by refactoring and using async/await.
Note, in the refactor flow there is a small UX change (an improvement IMO). There is now a flashMessage for each role whose transformation was not removed. Before, there would only be one generic flashMessage. Happy to wordsmith here; it's certainly not an easy concept to grasp.
image

2. The promise-proxy error also showed up anytime one of the `options-for-backend` dynamic components used the property `this.capabilities` directly inside a template—example [here](https://github.com/hashicorp/vault/blob/main/ui/app/templates/vault/cluster/secrets/backend/secret-edit-layout.hbs#L13). This is a tough workflow of logic, so bear with me:
* If you console logged `this.capabilities` in these templates, it returned a Proxy object.
image
* If you access this.capabilities (which for posterity is defined in secret-edit [here](https://github.com/hashicorp/vault/blob/f634808ed4b881af97df02b19eb15dd12c5051d2/ui/app/routes/vault/cluster/secrets/backend/secret-edit.js#L29-L44)) directly on a template you're directly accessing the proxy object. And this is a no-no. In all other use cases we are doing something to this object before accessing it on the template.

**Solution:**
For all of the use-cases that hit this I was able to confirm on the models they had used `lazyCapabilities` and either had getters on the models I could access or `updatePath` that I could replace `this.capabilities.canDelete` with `this.model.updatePath.canDelete`. Double checked these permissions worked as expected via smoke testing (see allll the screenshots at the end for posterity).

For any remaining `this.capabilities` in templates you can trace the original component manipulating this object (thus waiting first for it to resolve) in the js file.

**Note**: I was dodging rabbit holes right and left. So much refactoring can be done to these older models, flows, and the whole Transform engine should be redone. I jumped over the holes trying to keep scope focused as we have lots of work in front of us.

- [x] ent test pass locally

Screenshots below
* Transformation Edit
image
image

* Transit-form-edit
image
image

* Alphabet edit
image
image

*Transform-role-edit
image
image

*Transform-template-edit
image
image

*Transit form edit (permissions right, no permissions left)
image

*Transit form show (permissions right, no permissions left)
image

*Policy show (permissions right, no permissions left)
image

* Policy edit (permissions right, no permissions left)
image

### CODE REVIEW

File: ui/app/components/transformation-edit.js [hunk 1]
```diff
@@ -17,30 +17,12 @@ export default TransformBase.extend({
...
   },
 
-  updateOrCreateRole(role, transformationId, backend) {
-    return this.store
+  async updateOrCreateRole(role, transformationId, backend) {
+    const roleStore = await this.store
       .queryRecord('transform/role', {
         backend,
         id: role.id,
       })
```
Comment (Monkeychip) [inline ui/app/components/transformation-edit.js:25]: Just removed the .then part here and continue with the code flow as was before.

File: ui/app/components/alphabet-edit.hbs
Comment (Monkeychip) [inline ui/app/components/alphabet-edit.hbs:1]: Smoke test:
image
With policy:
image

File: ui/app/templates/components/transform-role-edit.hbs
Comment (Monkeychip) [inline ui/app/templates/components/transform-role-edit.hbs:1]: Smoke test:
image
image

File: ui/app/templates/components/transform-template-edit.hbs
Comment (Monkeychip) [inline ui/app/templates/components/transform-template-edit.hbs:1]: Smoke test:
image
image

File: ui/app/templates/components/transformation-edit.hbs
Comment (Monkeychip) [inline ui/app/templates/components/transformation-edit.hbs:1]: Smoke test:
image
image

File: ui/app/templates/components/transit-form-edit.hbs
Comment (Monkeychip) [inline ui/app/templates/components/transit-form-edit.hbs:1]: No comment about us hiding buttons here. Just focusing on the deprecations. transit needs lots of love. 
image
image

File: ui/app/components/alphabet-edit.hbs
Comment (Monkeychip) [inline ui/app/components/alphabet-edit.hbs:1]: Smoke test:
image
With policy:
image

File: ui/app/templates/components/transform-role-edit.hbs
Comment (Monkeychip) [inline ui/app/templates/components/transform-role-edit.hbs:1]: Smoke test:
image
image

File: ui/app/templates/components/transform-template-edit.hbs
Comment (Monkeychip) [inline ui/app/templates/components/transform-template-edit.hbs:1]: Smoke test:
image
image

File: ui/app/templates/components/transit-form-edit.hbs
Comment (Monkeychip) [inline ui/app/templates/components/transit-form-edit.hbs:1]: Avoiding a huge rabbit hole, ignore the giant blue button. Permissions on right, no permissions on left.
image

File: ui/app/templates/components/transit-form-show.hbs
Comment (Monkeychip) [inline ui/app/templates/components/transit-form-show.hbs:1]: Permissions on right, no permissions on left.
image

File: ui/app/templates/components/transit-edit.hbs
Comment (Monkeychip) [inline ui/app/templates/components/transit-edit.hbs:1]: these child components both use the models they're passed to access permissions instead of passing in the capabilities service (which was a strange pattern anyways—i confirmed, transit-edit does not create/edit a param called capabilities). Note too transit-edit is a component generated and rendered via the options-for-backend; here's the snipped so you can see all the things it has access to.
```
 transit: {
    searchPlaceholder: 'Filter keys',
    item: 'key',
    create: 'Create key',
    navigateTree: false,
    editComponent: 'transit-edit',
    listItemPartial: 'secret-list/item',
    firstStep: `To use transit, you'll need to create a key`,
  }
```

File: ui/app/templates/components/transit-form-edit.hbs
Comment (Monkeychip) [inline ui/app/templates/components/transit-form-edit.hbs:1]: my guess as to why were were checking @key.canDelete and then permissions on capabilities is because we had a one-off poorly named getter on the key model 'canDelete' that check if the key had the property allowDelete set to true.

File: ui/app/components/transformation-edit.js [hunk 2]
```diff
@@ -64,23 +46,39 @@ export default TransformBase.extend({
...
+    const { backend } = this.model;
+    updateRoles.forEach(async (record) => {
+      // For each role that needs to be updated, update the role with the transformation.
+      const updateOrCreateResponse = await this.updateOrCreateRole(record, transformationId, backend);
+      // If an error was returned, check error type and show a message.
+      const errorStatus = updateOrCreateResponse.errorStatus;
+      if (errorStatus) {
+        let message = `The edits to this transformation were successful, but transformations for the role ${record.id} were not edited due to a lack of permissions.`;
+        if (errorStatus !== 403) {
+          message = `You've edited the allowed_roles for this transformation. However, the corresponding edits to remove the role ${record.id}'s transformation was not made.`;
```
Comment (hellobontempo) [inline ui/app/components/transformation-edit.js:81]: I find it easier to put the dynamic string at the end (or in quotes) so it's clear(er) that it's interpolated. Also are we certain the error here will always be fore removing a role? I hit the error because of a problem with the role I attempted to add (this whole section is bad UX 😓 )
 
```suggestion
          message = `You've edited the allowed_roles for this transformation. However, there was a problem updating the role: ${record.id}.`;
```

File: ui/app/components/transformation-edit.js [hunk 3]
```diff
@@ -64,23 +46,39 @@ export default TransformBase.extend({
...
+    if (role.action === 'ADD') {
+      transformations = addToList(transformations, transformationId);
+    } else if (role.action === 'REMOVE') {
+      transformations = removeFromList(transformations, transformationId);
+    }
+    roleStore.setProperties({
+      backend,
+      transformations,
+    });
+    return roleStore.save().catch((e) => {
```
Comment (hellobontempo) [inline ui/app/components/transformation-edit.js:62]: It feels strange to have two catch blocks here - why did you originally remove the `.then`?

File: ui/app/components/transformation-edit.js [hunk 4]
```diff
@@ -64,23 +46,39 @@ export default TransformBase.extend({
...
-        if (results.find((e) => !!e.errorStatus && e.errorStatus !== 403)) {
-          // if the errors weren't all due to permissions show generic message
-          // eg. trying to update a role with empty array as transformations
-          message = `You've edited the allowed_roles for this transformation. However, the corresponding edits to some roles' transformations were not made`;
+    const { backend } = this.model;
+    updateRoles.forEach(async (record) => {
+      // For each role that needs to be updated, update the role with the transformation.
+      const updateOrCreateResponse = await this.updateOrCreateRole(record, transformationId, backend);
+      // If an error was returned, check error type and show a message.
+      const errorStatus = updateOrCreateResponse.errorStatus;
```
Comment (hellobontempo) [inline ui/app/components/transformation-edit.js:77]: should this be conditional in case there's no `errorStatus` object?
```suggestion
      const errorStatus = updateOrCreateResponse?.errorStatus;
```

File: ui/app/components/transformation-edit.js [hunk 5]
```diff
@@ -64,23 +46,39 @@ export default TransformBase.extend({
...
+      if (errorStatus) {
+        let message = `The edits to this transformation were successful, but transformations for the role ${record.id} were not edited due to a lack of permissions.`;
+        if (errorStatus !== 403) {
```
Comment (hellobontempo) [inline ui/app/components/transformation-edit.js:80]: I think this logic would be a little clearer without the nested conditional and explicitly checking for permissions first:
```suggestion
      if (errorStatus === 403) {
        let message = `The edits to this transformation were successful, but transformations for the role ${record.id} were not edited due to a lack of permissions.`;
       } else if (errorStatus) {
```

File: ui/app/templates/components/transform-role-edit.hbs [hunk 6]
```diff
@@ -30,7 +30,7 @@
 {{#if (eq this.mode "show")}}
   <Toolbar>
     <ToolbarActions>
-      {{#if this.capabilities.canDelete}}
+      {{#if this.model.updatePath.canDelete}}
```
Comment (hellobontempo) [inline ui/app/templates/components/transform-role-edit.hbs:33]: Is there a reason we don't want to make this a computed on the model? I ask because our new pattern has been to explicitly check for false so we don't accidentally hide links while the return is `undefined`. I'll defer to you, just wanted to call out!

File: ui/app/templates/components/transit-form-show.hbs [hunk 7]
```diff
@@ -63,7 +63,7 @@
         />
       {{/if}}
       {{#if (eq @mode "show")}}
-        {{#if (or @capabilities.canUpdate @capabilities.canDelete)}}
```
Comment (hellobontempo) [inline ui/app/templates/components/transit-form-show.hbs:66]: I don't even see a `@capabilities` arg passed into this component (and above)!? How the heck was this working before. I only looked because I was going to suggest cleaning up the unused args...I'm baffled 🤯

File: ui/app/templates/components/transit-edit.hbs [hunk 8]
```diff
@@ -40,14 +40,12 @@
     @deleteKey={{action "deleteKey"}}
     @key={{this.key}}
     @requestInFlight={{this.requestInFlight}}
-    @capabilities={{this.capabilities}}
```
Comment (hellobontempo) [inline ui/app/templates/components/transit-edit.hbs:43]: Thank you for remembering this cleanup! 🏅

File: ui/app/components/transformation-edit.js [hunk 9]
```diff
@@ -64,23 +46,39 @@ export default TransformBase.extend({
...
+      };
+    });
   },
 
   handleUpdateRoles(updateRoles, transformationId) {
     if (!updateRoles) return;
-    const backend = this.model.backend;
-    const promises = updateRoles.map((r) => this.updateOrCreateRole(r, transformationId, backend));
-
-    Promise.all(promises).then((results) => {
```
Comment (hellobontempo) [inline ui/app/components/transformation-edit.js:74]: Another way to do this is is use RSVP hash if we wanted to keep this logic flow, but I think the flash messages make sense for each role

File: ui/app/components/transformation-edit.js (hunk 3 above)
Comment (Monkeychip) [reply ui/app/components/transformation-edit.js:62]: Good question.

We returned the results from a .then() which returns a [new Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then). This was the main gist of the problem because this method gets called ([here](https://github.com/hashicorp/vault/blob/main/ui/app/components/transformation-edit.js#L72) is a link to the original code) on row 75. We then used then created a new array of new Promises, used the `all` to resolve all of them and moved on to error handling—but that was the gist of the error. Using unresolved promises. 

Using await returns a resolved promise, solving the problem. From there I went back and forth on how much to restructure this method based on that change. I opted to keep as much of the old structure as possible to avoid any regressions (and also because I thought it might be easier to read my changes in a review 🙃). That being said, if you think it no longer reads well I can revisit and make the new flow match more of async/await pattern instead of a patched async/await pattern. I could also maybe approach with a finally—something I didn't try.

File: ui/app/templates/components/transform-role-edit.hbs (hunk 6 above)
Comment (Monkeychip) [reply ui/app/templates/components/transform-role-edit.hbs:33]: I went that method at first, and hit some gnarly model isDestroyed during transitioning problems. I avoided that rabbit hole and stuck with this.

File: ui/app/components/transformation-edit.js (hunk 5 above)
Comment (Monkeychip) [reply ui/app/components/transformation-edit.js:80]: Agreed. I defaulted to using the old flow (maybe a little too much).

File: ui/app/components/transformation-edit.js (hunk 2 above)
Comment (Monkeychip) [reply ui/app/components/transformation-edit.js:81]: You're message is better, covering more use cases—thanks.

File: ui/app/templates/vault/cluster/policy/show.hbs [hunk 10]
```diff
@@ -36,7 +36,7 @@
       @extension={{if (eq this.policyType "acl") this.model.format "sentinel"}}
       @text="Download policy"
     />
-    {{#if (and (not-eq this.model.id "root") (or this.capabilities.canUpdate this.capabilities.canDelete))}}
```
[the remark this answers is not in the document]
Comment (Monkeychip) [reply ui/app/templates/vault/cluster/policy/show.hbs:39]: Let me double check that. Thanks!

File: ui/app/components/alphabet-edit.hbs
Comment (Monkeychip) [inline ui/app/components/alphabet-edit.hbs:1]: For transit only, I was unable to add the canDelete, etc getters on the model. It borked due to a isDestroyed transition error. opted for this fix instead. this whole area needs ❤️‍🩹

File: ui/app/templates/components/transform-role-edit.hbs (hunk 6 above)
Comment (Monkeychip) [reply ui/app/templates/components/transform-role-edit.hbs:33]: Updated a comment [here.](https://github.com/hashicorp/vault/pull/28563/files#r1785379398)

File: ui/app/templates/vault/cluster/policy/show.hbs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Monkeychip) [reply ui/app/templates/vault/cluster/policy/show.hbs:39]: Updated.

File: ui/app/components/transformation-edit.js [hunk 11]
```diff
@@ -17,30 +17,12 @@ export default TransformBase.extend({
     this.set('initialRoles', this.model.allowed_roles);
   },
 
-  updateOrCreateRole(role, transformationId, backend) {
-    return this.store
+  async updateOrCreateRole(role, transformationId, backend) {
+    const roleStore = await this.store
```
Comment (hellobontempo) [inline ui/app/components/transformation-edit.js:21]: can we call this `roleRecord`? it wasn't clear to me below that this was the record I thought it was maybe the store or something 💭

File: ui/app/components/transformation-edit.js [hunk 12]
```diff
@@ -64,28 +46,51 @@ export default TransformBase.extend({
...
+          }
+        );
+      } else if (errorStatus) {
+        this.flashMessages.info(
+          `You've edited the allowed_roles for this transformation. However, there was a problem updating the role: ${record.id}.`,
+          {
+            sticky: true,
+            priority: 300,
+          }
+        );
```
Comment (hellobontempo) [inline ui/app/components/transformation-edit.js:93]: Since we're not returning here, I wonder if this makes double flash messages appear? I think we can keep it to `let message` so we're only calling this.flashMessages once

File: ui/app/templates/components/transit-form-show.hbs [hunk 13]
```diff
@@ -63,7 +63,7 @@
...
+        {{#if (or @model.canUpdate @model.canDelete)}}
```
Comment (hellobontempo) [inline ui/app/templates/components/transit-form-show.hbs:66]: should this be `@model` or `@key`? Sorry - hard to follow all the capabilities and where they trace back to!

**Review by hellobontempo — APPROVED** [review]

File: ui/app/components/transformation-edit.js (hunk 11 above)
Comment (Monkeychip) [reply ui/app/components/transformation-edit.js:21]: Same thought, I only kept it the same for consistency to compare.

File: ui/app/components/transformation-edit.js (hunk 12 above)
Comment (Monkeychip) [reply ui/app/components/transformation-edit.js:93]: The let cannot be defined inside the if statement and accessed in the else statement (hence the double flash message). But let me move the let outside the if all together and put flow back.

File: ui/app/templates/components/transit-form-show.hbs (hunk 13 above)
Comment (Monkeychip) [reply ui/app/templates/components/transit-form-show.hbs:66]: 100% I went back and forth and I'll double check just to be sure. the edit form uses key as the model, and while this could also use key, the param model is the standard.

Comment (Monkeychip) [reply ui/app/templates/components/transit-form-show.hbs:66]: [Here](https://github.com/hashicorp/vault/pull/28563/files#diff-d0e181623443a26de2e79ee21726bd42287d5ce9c625ce6bda10e8f68e596bc7R50) the model was passed in for the show form.
